js/sri4node-attachments.js

```diff
@@ -1,1078 +1,1677 @@
+    //       accept(data);
+    //     }
+    //   });
+    // });
+  }
 
-      await deleteFromS3([tmpFileName]);
+  /**
+   * Rename a file on S3
+   *
+   * @param {TMultiPartSingleBodyForFileUploadsWithFileObj} fileWithJson
+   * @returns {Promise<void>}
+   */
+  async function renameFile(fileWithJson) {
+    const { fileObj } = fileWithJson;
+    const s3filename = getS3FileNameByMuliPartBody(fileWithJson);
+    const { tmpFileName } = fileObj;
+    debug(`Rename ${tmpFileName} to ${s3filename}`);
+
+    await copyFile(s3filename, tmpFileName, fileWithJson.attachment.key);
+
+    await deleteFromS3([tmpFileName]);
+  }
 
-    }
+  /**
+   *
+   * @param {import('stream').Readable} fileStream
+   * @param {string} tmpFileName
+   * @returns {Promise<import("@aws-sdk/client-s3").CompleteMultipartUploadCommandOutput | import("@aws-sdk/client-s3").AbortMultipartUploadCommandOutput>}
+   */
+  async function handleFileUpload(fileStream, tmpFileName) {
+    const awss3 = getAWSS3Client();
+
+    debug(`Uploading file ${tmpFileName}`);
+    const params = {
+      Bucket: fullPluginConfig.s3bucket,
+      Key: tmpFileName,
+      ACL: "bucket-owner-full-control",
+      Body: fileStream,
+    }; // , Metadata: { "attachmentkey": fileWithJson.attachment.key }
+
+    return await new Upload({ client: awss3, params}).done();
+  }
 
-    async function copyFile(destionationFileName, sourceFileName, attachmentKey) {
-      let awss3 = createAWSS3Client();
-      let params = { Bucket: configuration.s3bucket, Key: destionationFileName, ACL: "bucket-owner-full-control", CopySource: encodeURI("/" + configuration.s3bucket + "/" + sourceFileName), MetadataDirective: "REPLACE", TaggingDirective: "COPY", Metadata: { "attachmentkey": attachmentKey } };
-
-      await new Promise((accept, reject) => {
-        awss3.copyObject(params, function (err, data) {
-          if (err) { // an error occurred
-            //console.log(err, err.stack)
-            reject(err);
-          } else {
-            //console.log(data); // successful response
-            accept(data);
-          }
+  /**
+   *
+   * @param {IDatabase} tx
+   * @param {TSriRequest} sriRequest
+   * @param {import('stream').Readable} stream
+   * @param {boolean} isRetry
+   */
+  async function handleFileDownload(tx, sriRequest, stream, isRetry) {
+    const { filename } = sriRequest.params;
+    const safeFilename = getSafeFilename(filename);
+    const remoteFilename = `${sriRequest.params.key}-${
+      isRetry ? filename : safeFilename
+    }`;
+    try {
+      await downloadFromS3(stream, remoteFilename);
+    } catch (err) {
+      if (err === 404) {
+        if (isRetry || filename === safeFilename) {
+          throw new sriRequest.SriError({
+            status: 404,
+            errors: [
+              {
+                code: "download.failed",
+                type: "ERROR",
+                message: `Unable to download file [${filename}]`,
```

Comment:
Since it's a 404 I'd expect the message and code to be different, something like `file not found`.